edward/inferences/conjugacy/conjugacy.py

```diff
@@ -0,0 +1,246 @@
+  g = tf.get_default_graph()
+  blanket_name = _log_joint_name(blanket)
+  c = g.get_collection(blanket_name)
+  if len(c):
+    return c[0]
+
+  with tf.name_scope('conjugate_log_joint') as scope:
+    terms = []
+    for b in blanket:
+      if getattr(b, "conjugate_log_prob", None) is None:
+        raise NotImplementedError("conjugate_log_prob not implemented for"
+                                  " {}".format(type(b)))
+      terms.append(tf.reduce_sum(b.conjugate_log_prob()))
+    result = tf.add_n(terms, name=scope)
+    g.add_to_collection(blanket_name, result)
+    return result
+
+
+def complete_conditional(rv, blanket, log_joint=None):
+  with tf.name_scope('complete_conditional_%s' % rv.name) as scope:
+    # log_joint holds all the information we need to get a conditional.
+    extended_blanket = copy(blanket)
+    if log_joint is None:
+      log_joint = get_log_joint(extended_blanket)
+    else:
+      log_joint = log_joint(extended_blanket)
+
+    # Pull out the nodes that are nonlinear functions of rv into s_stats.
+    stop_nodes = set([i.value() for i in blanket])
+    subgraph = extract_subgraph(log_joint, stop_nodes)
+    s_stats = suff_stat_nodes(subgraph, rv.value(), blanket)
+    s_stats = list(set(s_stats))
+
+    # Simplify those nodes, and extract any new linear terms into multipliers_i.
+    s_stat_exprs = defaultdict(list)
+    for i in range(len(s_stats)):
+      expr = symbolic_suff_stat(s_stats[i], rv.value(), stop_nodes)
+      expr = full_simplify(expr)
+      multipliers_i, s_stats_i = extract_s_stat_multipliers(expr)
+      s_stat_exprs[s_stats_i].append((s_stats[i],
+                                      reconstruct_multiplier(multipliers_i)))
+
+    # Sort out the sufficient statistics to identify this conditional's family.
+    s_stat_keys = list(s_stat_exprs.keys())
+    order = np.argsort([str(i) for i in s_stat_keys])
+    dist_key = tuple((s_stat_keys[i] for i in order))
+    dist_constructor, constructor_params = (
+        _suff_stat_to_dist[rv.support].get(dist_key, (None, None)))
+    if dist_constructor is None:
+      raise NotImplementedError('Conditional distribution has sufficient '
+                                'statistics %s, but no available '
+                                'exponential-family distribution has those '
+                                'sufficient statistics.' % str(dist_key))
+
+    # Swap sufficient statistics for placeholders, then take gradients
+    # w.r.t.  those placeholders to get natural parameters. The original
+    # nodes involving the sufficient statistic nodes are swapped for new
+    # nodes that depend linearly on the sufficient statistic placeholders.
+    s_stat_nodes = []
+    s_stat_replacements = []
+    s_stat_placeholders = []
+    swap_dict = {}
+    swap_back = {}
+    for s_stat in s_stat_exprs.keys():
+      s_stat_shape = s_stat_exprs[s_stat][0][0].get_shape()
+      s_stat_placeholder = tf.placeholder(np.float32, s_stat_shape)
```

Comment:
It's because we want to take the gradient w.r.t. the placeholders, and TF doesn't like taking gradients w.r.t. non-floats.